src/json-patch-duplex.ts

```diff
@@ -659,82 +816,93 @@ module jsonpatch {
       if (operation.op == "add") {
         var pathLen = operation.path.split("/").length;
         var existingPathLen = existingPathFragment.split("/").length;
         if (pathLen !== existingPathLen + 1 && pathLen !== existingPathLen) {
-          throw new JsonPatchError('Cannot perform an `add` operation at the desired path', 'OPERATION_PATH_CANNOT_ADD', index, operation, tree);
+          throw new JsonPatchError('Cannot perform an `add` operation at the desired path', 'OPERATION_PATH_CANNOT_ADD', index, operation, document);
         }
       }
       else if (operation.op === 'replace' || operation.op === 'remove' || (<any>operation.op) === '_get') {
         if (operation.path !== existingPathFragment) {
-          throw new JsonPatchError('Cannot perform the operation at a path that does not exist', 'OPERATION_PATH_UNRESOLVABLE', index, operation, tree);
+          throw new JsonPatchError('Cannot perform the operation at a path that does not exist', 'OPERATION_PATH_UNRESOLVABLE', index, operation, document);
         }
       }
       else if (operation.op === 'move' || operation.op === 'copy') {
         var existingValue: any = { op: "_get", path: operation.from, value: undefined };
-        var error = jsonpatch.validate([existingValue], tree);
+        var error = validate([existingValue], document);
         if (error && error.name === 'OPERATION_PATH_UNRESOLVABLE') {
-          throw new JsonPatchError('Cannot perform the operation from a path that does not exist', 'OPERATION_FROM_UNRESOLVABLE', index, operation, tree);
+          throw new JsonPatchError('Cannot perform the operation from a path that does not exist', 'OPERATION_FROM_UNRESOLVABLE', index, operation, document);
         }
       }
     }
   }
 
   /**
-   * Validates a sequence of operations. If `tree` parameter is provided, the sequence is additionally validated against the object tree.
+   * Validates a sequence of operations. If `document` parameter is provided, the sequence is additionally validated against the object document.
    * If error is encountered, returns a JsonPatchError object
-   * @param sequence
-   * @param tree
+   * @param sequence 
+   * @param document
    * @returns {JsonPatchError|undefined}
    */
-  export function validate(sequence: Patch<any>[], tree?: any): JsonPatchError {
+  export function validate<T>(sequence: Operation[], document?: T, externalValidator?: Validator<T>): JsonPatchError {
     try {
       if (!_isArray(sequence)) {
         throw new JsonPatchError('Patch sequence must be an array', 'SEQUENCE_NOT_AN_ARRAY');
       }
 
-      if (tree) {
-        tree = JSON.parse(JSON.stringify(tree)); //clone tree so that we can safely try applying operations
-        apply.call(this, tree, sequence, true);
+      if (document) {
+        document = JSON.parse(JSON.stringify(document)); //clone document so that we can safely try applying operations
+        if(typeof externalValidator == 'function') {
```

Comment:
can we just make it  `externalValidator = externalValidator || validator` and remove those two `if`s?